public/app/plugins/panel/geomap/layers/data/geojsonLayer.ts

```diff
@@ -131,6 +139,29 @@ export const geojsonLayer: MapLayerRegistryItem<GeoJSONMapperConfig> = {
             return polyStyle(values);
           }
 
+          // Support styling polygons from Feature properties
+          const featureProps = feature.getProperties();
+          if (isPolygon && Object.keys(featureProps).some((property) => polyStyleStrings.includes(property))) {
```

Comment:
I'm a bit concerned about the property comparisons here. Seems like there could be a situation in which we have M `featureProps` along with N style strings resulting in worse case M*N in terms of comparisons. Probably not meaningful in small layers, but I could see this being hairy in larger ones potentially.

Reply:
Thanks for testing it!

Perhaps we could do the embedded styling check for geojson on first load and store a file content hash with hasStyling boolean inside of the panel json. That way, if there is no embedded styling in a particular source (that contents hasn't changed), it avoids these loops altogether. Of course doing the hash check wouldn't be free and there may be a crossover point where it doesn't make sense. For the case where embedded styling is found, we could store a comprehensive map of which styles were used to reduce these loops.

TLDR: we should follow-up on how to better optimize this.

Reply:
That sounds like a really good way forward! Really appreciate the thoughtful approach here 😄